File: NEsper/eql/join/plan/FilterExprAnalyzer.cs
using System;

using net.esper.eql.expression;

namespace net.esper.eql.join.plan
{

    /// <summary> Analyzes a filter expression and builds a query graph model.
    /// The 'equals' and 'and' expressions in the filter expression are extracted
    /// and placed in the query graph model as navigable relationships (by key and index
    /// properties) between streams.
    /// </summary>
    public class FilterExprAnalyzer
    {
        /// <summary> Analyzes filter expression to build query graph model.</summary>
        /// <param name="topNode">filter top node
        /// </param>
        /// <param name="queryGraph">model containing relationships between streams, to be written to
        /// </param>
        public static void Analyze(ExprNode topNode, QueryGraph queryGraph)
        {
            // Analyze relationships between streams. Relationships are properties in AND and EQUALS nodes of joins.
            if (topNode is ExprEqualsNode)
            {
                ExprEqualsNode EqualsNode = (ExprEqualsNode)topNode;
                if (!EqualsNode.IsNotEquals)
                {
                    AnalyzeEqualsNode(EqualsNode, queryGraph);
                }
            }
            else if (topNode is ExprAndNode)
            {
                ExprAndNode andNode = (ExprAndNode)topNode;
                AnalyzeAndNode(andNode, queryGraph);
            }
        }

        /// <summary> Analye EQUALS (=) node.</summary>
        /// <param name="equalsNode">node to analyze
        /// </param>
        /// <param name="queryGraph">store relationships between stream properties
        /// </param>
        public static void AnalyzeEqualsNode(ExprEqualsNode equalsNode, QueryGraph queryGraph)
        {
            if ((!(equalsNode.ChildNodes[0] is ExprIdentNode)) || (!(equalsNode.ChildNodes[1] is ExprIdentNode)))
            {
                return;
            }

            ExprIdentNode identNodeLeft = (ExprIdentNode)equalsNode.ChildNodes[0];
            ExprIdentNode identNodeRight = (ExprIdentNode)equalsNode.ChildNodes[1];

            queryGraph.Add(identNodeLeft.StreamId, identNodeLeft.ResolvedPropertyName, identNodeRight.StreamId, identNodeRight.ResolvedPropertyName);
        }

        /// <summary> Analyze the AND-node.</summary>
        /// <param name="andNode">node to analyze
        /// </param>
        /// <param name="queryGraph">to store relationships between stream properties
        /// </param>
        public static void AnalyzeAndNode(ExprAndNode andNode, QueryGraph queryGraph)
        {
            foreach (ExprNode childNode in andNode.ChildNodes)
            {
                ExprEqualsNode equalsNode = childNode as ExprEqualsNode;
                if ((equalsNode != null) && (!equalsNode.IsNotEquals))
                {
                    AnalyzeEqualsNode(equalsNode, queryGraph);
                }
            }
        }
    }
}

File: trunk/esper/src/main/java/net/esper/eql/join/plan/FilterExprAnalyzer.java
/**************************************************************************************
 * Copyright (C) 2006 Esper Team. All rights reserved.                                *
 * http://esper.codehaus.org                                                          *
 * ---------------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the GPL license       *
 * a copy of which has been included with this distribution in the license.txt file.  *
 **************************************************************************************/
package net.esper.eql.join.plan;

import net.esper.eql.expression.*;
import net.esper.util.JavaClassHelper;

/**
 * Analyzes a filter expression and builds a query graph model.
 * The 'equals' and 'and' expressions in the filter expression are extracted
 * and placed in the query graph model as navigable relationships (by key and index
 * properties) between streams.
 */
public class FilterExprAnalyzer
{
    /**
     * Analyzes filter expression to build query graph model.
     * @param topNode - filter top node
     * @param queryGraph - model containing relationships between streams, to be written to
     * @param isAllowCoercion - true to indicate that coercing filter criteria are considered, false if not
     */
    public static void analyze(ExprNode topNode, QueryGraph queryGraph, boolean isAllowCoercion)
    {
        // Analyze relationships between streams. Relationships are properties in AND and EQUALS nodes of joins.
        if (topNode instanceof ExprEqualsNode)
        {
            ExprEqualsNode equalsNode = (ExprEqualsNode) topNode;
            if (!equalsNode.isNotEquals())
            {
                analyzeEqualsNode(equalsNode, queryGraph, isAllowCoercion);
            }
        }
        else if (topNode instanceof ExprAndNode)
        {
            ExprAndNode andNode = (ExprAndNode) topNode;
            analyzeAndNode(andNode, queryGraph, isAllowCoercion);
        }
    }

    /**
     * Analye EQUALS (=) node.
     * @param equalsNode - node to analyze
     * @param queryGraph - store relationships between stream properties
     * @param isAllowCoercion - true to indicate that coercing filter criteria are considered, false if not
     */
    protected static void analyzeEqualsNode(ExprEqualsNode equalsNode, QueryGraph queryGraph, boolean isAllowCoercion)
    {
        if ( (!(equalsNode.getChildNodes().get(0) instanceof ExprIdentNode)) ||
             (!(equalsNode.getChildNodes().get(1) instanceof ExprIdentNode)) )
        {
            return;
        }

        ExprIdentNode identNodeLeft = (ExprIdentNode) equalsNode.getChildNodes().get(0);
        ExprIdentNode identNodeRight = (ExprIdentNode) equalsNode.getChildNodes().get(1);

        if (!isAllowCoercion)
        {
            try
            {
                Class boxedTypeLeft = JavaClassHelper.getBoxedType(identNodeLeft.getType());
                Class boxedTypeRight = JavaClassHelper.getBoxedType(identNodeRight.getType());
                if (!(boxedTypeLeft.equals(boxedTypeRight)))
                {
                    return;
                }
            }
            catch (ExprValidationException e)
            {
                throw new IllegalStateException(e.getMessage(), e);
            }
        }

        queryGraph.add(identNodeLeft.getStreamId(), identNodeLeft.getResolvedPropertyName(),
                identNodeRight.getStreamId(), identNodeRight.getResolvedPropertyName());
    }

    /**
     * Analyze the AND-node.
     * @param andNode - node to analyze
     * @param queryGraph - to store relationships between stream properties
     * @param isAllowCoercion - true to indicate that coercing filter criteria are considered, false if not
     */
    protected static void analyzeAndNode(ExprAndNode andNode, QueryGraph queryGraph, boolean isAllowCoercion)
    {
        for (ExprNode childNode : andNode.getChildNodes())
        {
            if (childNode instanceof ExprEqualsNode)
            {
                ExprEqualsNode equalsNode = (ExprEqualsNode) childNode;
                if (!equalsNode.isNotEquals())
                {
                    analyzeEqualsNode(equalsNode, queryGraph, isAllowCoercion);
                }
            }
        }
    }
}
